File: leet-code/167. Two Sum II - Input Array Is Sorted/TwoSumInputArrayIsSorted167_2.java
 package com.example;

public class TwoSumInputArrayIsSorted167_2 {
    public int[] twoSum(int[] numbers, int target) {
        int[] answer = new int[2];
        for (int i = 0; i < numbers.length - 1; i++) {
            int num = numbers[i];
            int targetIdx = getTargetIdx(numbers, i + 1, numbers.length - 1, target - num);
            if (targetIdx == -1) continue;
            answer[0] = i + 1;
            answer[1] = targetIdx + 1;
        }
        return answer;
    }

    private int getTargetIdx(int[] numbers, int left, int right, int target) {
        while (left <= right) {
            int mid = left + (right - left) / 2;
            if (numbers[mid] == target) return mid;
            if (numbers[mid] < target) left = mid + 1;
            else right = mid - 1;
        }
        return -1;
    }
}

File: leet-code/167. Two Sum II - Input Array Is Sorted/TwoSumInputArrayIsSorted167_2.kt
package com.example

class TwoSumInputArrayIsSorted167_2 {
    fun twoSum(numbers: IntArray, target: Int): IntArray {
        val answer = IntArray(2)
        for (idx1 in 0..numbers.size - 2) {
            val idx2 = getTarget(numbers, target, numbers[idx1], idx1)
            if (idx2 == 0) {
                continue
            }
            answer[0] = idx1 + 1
            answer[1] = idx2 + 1
            break
        }
        return answer
    }

    private fun getTarget(numbers: IntArray, target: Int, num1: Int, idx: Int): Int {
        var low = idx + 1
        var high = numbers.size - 1
        while (low <= high) {
            val mid = low + (high - low) / 2
            val num2 = numbers[mid]
            if (num1 + num2 == target) {
                return mid
            }
            if (num1 + num2 < target) {
                low = mid + 1
            } else {
                high = mid - 1
            }
        }
        return 0
    }
}